funderfinder/utils/list_gsoc.py

```diff
@@ -0,0 +1,263 @@
+    :param link_container: BeautifulSoup container that may contain GSOC link
+    :return: Cleaned GSOC link, if link was found
+    """
+    link = link_container.find("a", href=True)
+    if link:
+        return "https://www.google-melange.com" + link["href"]
+    return None
+
+
+def get_early_archive_listing_links(soup: bs4.BeautifulSoup) -> list:
+    """
+    Extract listing link containers from an old-format GSOC page. These currently always
+    appear in spans with the class shown below
+    :param soup: BeautifulSoup element containing GSOC links
+    :return: A list of BeautifulSoup elements corresponding to GSOC link containers
+    """
+    return soup.find_all("span", {"class": "mdl-list__item-primary-content"})
+
+
+def get_link_matches(text: str) -> list:
+    """
+    Extract any github orgs (if not repo specified) or repos in text
+    :param text: Text that may contain github links
+    :return: List of orgs and/or repos
+    """
+    repo_matches = re.findall(GITHUB_REPO_PATTERN, text)
+    if repo_matches:
+        return list(set([repo_match[1] for repo_match in repo_matches]))
+    # At the moment, we only look for org matches if we didn't find a specific repo in the text
+    org_matches = re.findall(GITHUB_ORG_PATTERN, text)
+    if org_matches:
+        return list(set([org_match[1] for org_match in org_matches]))
+    return []
+
+
+def get_early_archive_repos(link: str) -> list:
+    """
+    Extracts repos/orgs for a pre-2016 project.
+    :param link: Link to GSOC's detail page
+    :return: List of github repos associated with the project
+    """
+    project_page = requests.get(link)
+    soup = bs4.BeautifulSoup(project_page.text, features="html.parser")
+    links = [
+        extract_listing_link(link_elt) for link_elt in soup.find_all("a", href=True)
+    ]
+    repos = []
+    for link in links:
+        if not link:
+            continue
+        gh_links = get_link_matches(link)
+        repos.extend(gh_links)
+    page_link = get_link_matches(project_page.text)
+    repos.extend(page_link)
+    student_projects = get_early_archive_listing_links(soup)
+    student_project_links = [
+        extract_listing_link(link_container) for link_container in student_projects
+    ]
+    for student_project_link in student_project_links:
+        if not student_project_link:
+            continue
+        student_project_page = requests.get(student_project_link)
+        student_page_links = get_link_matches(student_project_page.text)
+        if student_page_links:
+            repos.extend(student_page_links)
+        time.sleep(SCRAPE_DELAY)
+    return list(set(repos))
+
+
+def get_early_archive_project(container: bs4.BeautifulSoup, year: int) -> dict:
+    """
+    Get a pre-2016 project's metadata
+    :param container: BeautifulSoup container containing the project's name and link to GSOC's detail page
+    :param year: Year the project appeared (a project may appear in more than one year)
```

Comment:
I don't yet understand this `Year the project appeared` comment. Is this the *first* year the project appeared? Something else?

Reply:
Good point, that is confusing - it's the year of GSOC that we're currently parsing. I've updated the comment to hopefully make this clearer!